precompile/assetsbridge/bridge_out.go

```diff
@@ -0,0 +1,615 @@
+	sdkCtx sdk.Context,
+	spenderAddr common.Address,
+	token common.Address,
+	amount *big.Int,
+) error {
+	// we use the ERC20 allowance call for either the BTC token
+	// or the ERC20 as both implements the ERC20 interface
+
+	bridgeAddrBytes := common.HexToAddress(
+		evmtypes.AssetsBridgePrecompileAddress,
+	).Bytes()
+
+	call, err := evmtypes.NewERC20AllowanceCall(
+		bridgeAddrBytes,
+		token.Bytes(),
+		spenderAddr.Bytes(),
+		// spendable by the asset bridge
+		common.HexToAddress(
+			evmtypes.AssetsBridgePrecompileAddress,
+		).Bytes(),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create ERC20 allowance call: %w", err)
+	}
+
+	resp, err := m.evmKeeper.ExecuteContractCall(sdkCtx, call)
+	if err != nil {
+		return fmt.Errorf("failed to execute ERC20 allowance call: %w", err)
+	}
+
+	// this would be non empty if the call reverted
+	if len(resp.VmError) > 0 {
+		// TODO(Jeremy): Should we handle the revert properly
+		// to return an error message? For now just returning
+		// a generic error message
+		return fmt.Errorf("ERC20 allowance call reverted")
+	}
+
+	allowance, err := extractAllowanceFromEVMResult(resp.Ret)
+	if err != nil {
+		return fmt.Errorf("unable to unpack EVM result: %v", err)
+	}
+
+	// if the amount <= to the balance, then OK
+	// else we do not have enough funds and we need
+	// to return an error
+	if amount.Cmp(allowance) > 0 {
+		return fmt.Errorf("asset bridge allowance too low")
+	}
+
+	return nil
+}
+
+func (m *BridgeOutMethod) isAmountAvailable(
+	sdkCtx sdk.Context,
+	spenderAddr common.Address,
+	token common.Address,
+	amount *big.Int,
+) error {
+	// we use the ERC20BalanceOf call for either the BTC token
+	// or the ERC20 as both implements the ERC20 interface
+
+	bridgeAddrBytes := common.HexToAddress(
+		evmtypes.AssetsBridgePrecompileAddress,
+	).Bytes()
+
+	call, err := evmtypes.NewERC20BalanceOfCall(
+		bridgeAddrBytes,
+		token.Bytes(),
+		spenderAddr.Bytes(),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create ERC20 balanceOf call: %w", err)
+	}
+
+	resp, err := m.evmKeeper.ExecuteContractCall(sdkCtx, call)
+	if err != nil {
+		return fmt.Errorf("failed to execute ERC20 balanceOf call: %w", err)
+	}
+
+	// this would be non empty if the call reverted
+	if len(resp.VmError) > 0 {
+		// TODO(Jeremy): Should we handle the revert properly
+		// to return an error message? For now just returning
+		// a generic error message
+		return fmt.Errorf("ERC20 balanceOf call reverted")
+	}
+
+	balance, err := extractBalanceOfFromEVMResult(resp.Ret)
+	if err != nil {
+		return fmt.Errorf("unable to unpack EVM result: %v", err)
+	}
+
+	// if the amount <= to the balance, then OK
+	// else we do not have enough funds and we need
+	// to return an error
+	if amount.Cmp(balance) > 0 {
+		return fmt.Errorf("not enough funds to bridgeOut")
+	}
+
+	return nil
+}
+
+func (m *BridgeOutMethod) isValidToken(sdkCtx sdk.Context, token common.Address) bool {
```

Comment:
This function misses the fact that tokens must be validated in the context of the chain (see https://linear.app/thesis-co/document/technical-architecture-9fbdf2ac111c#heading-validation-logic-16c41930). TLDR:
- If Ethereum: BTC and enabled ERC20 are valid tokens
- If Bitcoin: Only BTC is valid

Reply:
fixed: 5516f26e

Reply:
very likely, this should do: https://github.com/mezo-org/mezod/pull/509/commits/36e365dadef62b890227d618604c9df45fba7c72